### ISSUE TITLE
Firmware handoff library hosting place

### ISSUE THREAD

**Issue by manish-pandey-arm** [FirmwareHandoff/firmware_handoff#20]
I want to start this thread to discuss about hosting firmware handoff library. 

The initial implementation of the library has been posted in TF-A[1] as well as in OP-TEE [2] projects by raymond.mao@linaro.org . There is also plan to implement it in U-boot project.

Considering different FW projects are adopting it, should we think of hosting this as library and all core developments can happen directly here instead of modifying various projects. This also helps us in avoiding divergence in future.

[1] https://review.trustedfirmware.org/c/TF-A/trusted-firmware-a/+/22215
[2] https://github.com/OP-TEE/optee_os/pull/6308

**Comment by sjg20** [comment]
U-Boot's implementation (where this idea originally came from) is called bloblist.

There are some patches to update it for the smaller headers, but I can't link to them as patchwork is half dead ATM

**Comment by sjg20** [comment]
OK it came back, so the link is here:

https://patchwork.ozlabs.org/user/todo/uboot/?series=365719

I actually did a dual license thinking it might help to avoid too much duplication

**Comment by danh-arm** [comment]
I agree a common library implementation would be good. If we made this a common support library with a BSD-3-Clause license at tf.org, then we could do this without much extra Arm internal or tf.org approval. That license should also be usable by U-Boot. A dual licensed version is possible too but will require additional approval steps. I don't have a patchwork account but this is not really suitable upstream for a project with multiple contributors. Also, we should be careful about IP pollution until the license is agreed.

**Comment by jwerner-chromium** [comment]
I'm not against providing a standard library implementation for those who want to use it, but I think it would important to clarify that this remains a data format specification that any project can implement on their own if they want to. There should be no implied assumption that you need to use the library to remain fully compatible, and we should make sure that all data format definitions and changes that are relevant for interoperability are always clearly explained in the paper spec, not just implemented in the library. (I would be okay with using C headers from the library as source of truth for tag format definitions, though, if that's something that we wanted to do, because I feel the GitHub markdown is quickly getting pretty unwieldy to define all these structures. We just need to make sure we draw a clear line in the source repository between the headers that are officially part of the spec and the code that just provides the example implementation.)

For hosting I would also prefer the TF.org Gerrit. GitHub has a terrible review experience.

**Comment by raymo200915** [comment]
Currently the Transfer List codes in TF-A and OP-TEE are very close, so it will not be an issue for me to bring them to a standard library. However, U-Boot is using a way to be compatible with existing bloblist and I guess that needs some efforts to bring it to the ones that now is used in TF-A/OP-TEE. @sjg20 Please correct me if not the case.

**Comment by apalos** [comment]
I am fine with the library idea. As far as I am aware U-Boot and OP-TEE should have no problems pulling it and compiling it

**Comment by sjg20** [comment]
There is a series to update U-Boot, here:

U-Boot impl](https://patchwork.ozlabs.org/user/todo/uboot/?series=365719

But it needs another spin now that the header-size changes has landed. I will take a look. It is dual-licensed so could perhaps provide a reference impl, except for the init code which is U-Boot-specific

**Comment by raymo200915** [comment]
@sjg20 @apalos I would prefer to refactor U-Boot Bloblist to be a layer on top of Transfer List APIs which are currently implemented for TF-A/OP-TEE, so that we can put the most generic stuffs as a standalone library.

**Comment by sjg20** [comment]
You wrote a new library? Do you have a link to it?

**Comment by sjg20** [comment]
Hmmm, but why not base it on the U-Boot code?

**Comment by raymo200915** [comment]
> Hmmm, but why not base it on the U-Boot code?
@sjg20 
I started this in early of May of this year.
At that time I needed a quick solution to test with TF-A/OP-TEE and didn't know you were planing to do this with Bloblist.

**Comment by sjg20** [comment]
Oh I see. Actually this whole spec came from bloblist, just the name changed...but then we changed the headers too!

I do wonder whether it would be better to use the U-Boot code? It has tests and would save me the effort of implementing resize, etc.

**Comment by manish-pandey-arm** [comment]
I assume the agreement (in one of the Firmware handoff meetings) was to go ahead with a standalone library hosted at https://www.trustedfirmware.org/ along with unit tests and transfer list generation tool.

Looking at the past comments, i think it should be clearly mentioned that this is just a reference implementation and there is no implied assumption that you need to use the library to remain fully compatible.

This issue can be closed if people on above.

**Comment by manish-pandey-arm** [comment]
The library will be pushed on https://www.trustedfirmware.org/
